content/blog/2025-12-16-gvn-final-report.md

```diff
@@ -0,0 +1,1350 @@
+- **Dynamic Instruction Count**: Instructions executed at runtime (performance)
+- **Geometric Mean Ratio**: GM of (optimized/original) across benchmarks
+
+### 4.2 Results Summary
+
+| Metric | Value |
+|--------|-------|
+| Total Benchmarks | 123 |
+| All Tests Pass | 100% |
+| Static Instruction Ratio (GM) | **0.8411** (15.9% reduction) |
+| Dynamic Instruction Ratio (GM) | **0.8170** (18.3% reduction) |
+| Benchmarks with Static Improvement | 62/123 (50.4%) |
+| Benchmarks with Dynamic Improvement | 62/123 (50.4%) |
+| Benchmarks worse in both metrics | 22/123 (17.9%) |
+
+### 4.3 Best Improvements
+
+**Top 5 Static Improvements:**
+
+| Benchmark | Original | Optimized | Ratio |
+|-----------|----------|-----------|-------|
+| ray-sphere-intersection.bril | 106 | 30 | 0.283x |
+| pascals-row.bril | 41 | 18 | 0.439x |
+| euclid.bril | 51 | 23 | 0.451x |
+| sqrt_bin_search.bril | 44 | 20 | 0.455x |
+| sum-sq-diff.bril | 66 | 30 | 0.455x |
+
+**Top 5 Dynamic Improvements:**
+
+| Benchmark | Original | Optimized | Ratio |
+|-----------|----------|-----------|-------|
+| ray-sphere-intersection.bril | 142 | 42 | 0.296x |
+| sum-sq-diff.bril | 3038 | 1117 | 0.368x |
+| sqrt_bin_search.bril | 744 | 288 | 0.387x |
+| birthday.bril | 484 | 188 | 0.388x |
+| pascals-row.bril | 146 | 57 | 0.390x |
+
+### 4.4 Analysis by Category
+
+| Category | Benchmarks | Static GM | Dynamic GM |
+|----------|------------|-----------|------------|
+| core | 68 | 0.8154 | 0.7879 |
+| float | 18 | 0.7371 | 0.6948 |
+| mem | 31 | 0.9506 | 0.9591 |
+| mixed | 4 | 0.9728 | 0.9527 |
+| long | 2 | 0.8870 | 0.7373 |
+
+The `float` benchmarks show the best improvement (26.3% static, 30.5% dynamic) due to floating-point computations having many redundant expressions. The `core` benchmarks also show strong improvements (18.5% static, 21.2% dynamic). Memory-heavy benchmarks (`mem`) show minimal improvement because memory operations are effectful and cannot be optimized by GVN.
+
+### 4.5 Comparison with Get/Set SSA Implementation
+
+We compared our real phi node implementation against an alternative approach using Bril's `get`/`set` label-based syntax. While Bril supports `get`/`set` for SSA representation, the Cooper/Briggs paper assumes real phi nodes. We initially implemented GVN using `get`/`set`, then re-implemented it with real phi nodes for comparison.
+
+| Metric | Our Implementation (Real Phis) | Get/Set Implementation | Improvement |
+|--------|--------------------------------|------------------------|-------------|
+| Total Static (62 common tests) | 2,262 | 2,313 | +51 instructions better |
+| Static Reduction | 12.0% | 10.0% | +2.0% better |
+
+The final phi node implementation outperforms the `get`/`set` approach by 51 instructions on the common benchmark set, primarily because real phis expose phi node structure to the GVN algorithm, enabling better redundancy detection and more precise dominator-tree analysis.
```

Comment:
I don't immediately see why this is the case. Can you present a little bit of an argument to justify your claim?